diff --git a/CMakeLists.txt b/CMakeLists.txt
index 067657ce8..2ae564731 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -51,6 +51,8 @@ endif()
 ###############################################################################
 # Set up options
 
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -latomic -march=native -DBOOST_OPTIONAL_CONFIG_ALLOW_BINDING_TO_RVALUES")
+
 # Configurable Options
 if(GTSAM_UNSTABLE_AVAILABLE)
     option(GTSAM_BUILD_UNSTABLE              "Enable/Disable libgtsam_unstable"          ON)
