diff --git a/loop_fusion/CMakeLists.txt b/loop_fusion/CMakeLists.txt
index 29568f0..762860a 100644
--- a/loop_fusion/CMakeLists.txt
+++ b/loop_fusion/CMakeLists.txt
@@ -23,7 +23,12 @@ find_package(Ceres REQUIRED)
 set(CMAKE_MODULE_PATH ${PROJECT_SOURCE_DIR}/cmake)
 find_package(Eigen3)
 
-include_directories(${catkin_INCLUDE_DIRS} ${CERES_INCLUDE_DIRS}  ${EIGEN3_INCLUDE_DIR})
+include_directories(
+  ${catkin_INCLUDE_DIRS}
+  ${CERES_INCLUDE_DIRS}
+  ${EIGEN3_INCLUDE_DIR}
+  ../camera_models/include
+  )
 
 catkin_package()
 
