diff --git a/CMakeLists.txt b/CMakeLists.txt
index 6ce583bc6..865ee5d99 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -193,6 +193,20 @@ if(UNIX AND NOT ANDROID)
       list(APPEND CMAKE_LIBRARY_PATH /usr/lib)
     endif()
   endif()
+  if(ARM OR AARCH64)
+    if(EXISTS /usr/lib/arm-linux-gnueabihf)
+      list(APPEND CMAKE_LIBRARY_PATH /usr/lib/arm-linux-gnueabihf)
+    endif()
+    if(EXISTS /usr/include/arm-linux-gnueabihf)
+      list(APPEND CMAKE_INCLUDE_PATH /usr/include/arm-linux-gnueabihf)
+    endif()
+  endif()
+  if(EXISTS /usr/include/linux)
+    list(APPEND CMAKE_INCLUDE_PATH /usr/include/linux)
+  endif()
+  if(EXISTS /usr/lib/linux)
+    list(APPEND CMAKE_INCLUDE_PATH /usr/lib/linux)
+  endif()
 endif()
 
 # Add these standard paths to the search paths for FIND_PATH
diff --git a/cmake/OpenCVFindOpenBLAS.cmake b/cmake/OpenCVFindOpenBLAS.cmake
index ae2daaa19..5f2c2183e 100644
--- a/cmake/OpenCVFindOpenBLAS.cmake
+++ b/cmake/OpenCVFindOpenBLAS.cmake
@@ -53,6 +53,7 @@ SET(Open_BLAS_INCLUDE_SEARCH_PATHS
   /usr/include/openblas-base
   /usr/local/include
   /usr/include
+  /usr/include/arm-linux-gnueabihf
 )
 
 SET(Open_BLAS_LIB_SEARCH_PATHS
@@ -69,6 +70,7 @@ SET(Open_BLAS_LIB_SEARCH_PATHS
         /usr/lib/openblas-base
         /usr/lib64
         /usr/lib
+        /usr/lib/arm-linux-gnueabihf
  )
 
 FIND_PATH(OpenBLAS_INCLUDE_DIR NAMES cblas.h PATHS ${Open_BLAS_INCLUDE_SEARCH_PATHS} NO_DEFAULT_PATH)
