diff --git a/CMakeLists.txt b/CMakeLists.txt
index af7eebd2..09700ff4 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -85,6 +85,9 @@ IF(CMAKE_COMPILER_IS_GNUCXX)
   SET(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fmessage-length=0")
 ENDIF(CMAKE_COMPILER_IS_GNUCXX)
 
+SET(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -latomic")
+SET(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++11")
+
 if(MSVC)
   if(MSVC_VERSION GREATER 1500 AND ${CMAKE_VERSION} VERSION_GREATER "2.8.6")
     include(ProcessorCount)
@@ -198,6 +201,8 @@ endif()
 
 FIND_PACKAGE(ZLIB REQUIRED QUIET)
 
+FIND_PACKAGE(CSparse REQUIRED)
+
 FIND_PACKAGE(Sqlite3 QUIET)
 IF(Sqlite3_FOUND)
    MESSAGE(STATUS "Found Sqlite3: ${Sqlite3_INCLUDE_DIRS} ${Sqlite3_LIBRARIES}")
@@ -261,8 +266,8 @@ IF(WITH_QT)
         ELSE()
             list(FIND PCL_LIBRARIES vtkGUISupportQt value)
             IF(value EQUAL -1)
-                #SET(PCL_LIBRARIES "${PCL_LIBRARIES};vtkGUISupportQt")
-                #SET(ADD_VTK_GUI_SUPPORT_QT_TO_CONF TRUE)
+                SET(PCL_LIBRARIES "${PCL_LIBRARIES};vtkGUISupportQt")
+                SET(ADD_VTK_GUI_SUPPORT_QT_TO_CONF TRUE)
             ENDIF(value EQUAL -1)
 			MESSAGE(STATUS "VTK_RENDERING_BACKEND=${VTK_RENDERING_BACKEND}")
 			IF(VTK_RENDERING_BACKEND STREQUAL "OpenGL2")
